django/db/models/query.py

```diff
@@ -1909,6 +1910,8 @@ def _clone(self):
     def _fetch_all(self):
         if self._result_cache is None:
             self._result_cache = list(self._iterable_class(self))
+            self.executed_query = connections[self.db].last_query
```

Comment:
Thanks for the PR!

As mentioned during the Djangonauts Review session, this could be a problem if the queryset does not execute a database query at all e.g. `objects.none()`, as it would return the last executed query which may come from a different queryset.

A potential workaround is to get a snapshot of `connections[self.db].queries` before the iterable is evaluated. Then, compare the snapshot with what the list of queries are now. If there are new queries, then those are likely to be related to the current queryset. This might also help with prefetch queries. So maybe instead of `executed_query`, we should make it `executed_queries` instead which contains the new queries. This means we return an empty list if no queries are executed – which may also avoid having to decide whether to return `None` or raise an exception.

Of course, this solution might not be safe for race conditions. Though I'm not sure if the current approach is safe either 🤔

Reply:
I wonder if Django would be better off if there was a signal that was raised when queries are executed.

Reply:
It might. I think I should take this discussion to the forum, it doesn't belong on this particular PR. I do like your idea Sage of having an `executed_queries` as that seems more future proof.

Reply:
Another issue with using a signal is that it is possible to perform a query during a signal receiver causing an infinite loop of queries.

Reply:
I like @laymonage's idea of snapshoting. We could snapshot the length of `connection.queries` list before and after evaluation and have `executed_query` (or `executed_queries` assuming we want to include prefetching as well) be a slice of it.